bundles/org.eclipse.swt/Eclipse SWT Browser/win32/org/eclipse/swt/browser/Edge.java

```diff
@@ -282,6 +279,95 @@ static int callAndWait(String[] pstr, ToIntFunction<IUnknown> callable) {
 	return phr[0];
 }
 
+class WebViewProvider {
+
+	private CompletableFuture<ICoreWebView2> webViewFuture = new CompletableFuture<>();
+	private CompletableFuture<ICoreWebView2_2> webView_2Future = new CompletableFuture<>();
+	private CompletableFuture<ICoreWebView2_11> webView_11Future = new CompletableFuture<>();
+
+	private CompletableFuture<Void> lastWebViewTask = webViewFuture.thenCompose(__ -> {
+		return CompletableFuture.allOf(webView_2Future.handle((result, ex) -> null), webView_11Future.handle((result, ex) -> null));
+	});
+
+
+	private ICoreWebView2 getWebView(boolean waitForPendingWebviewTasksToFinish) {
+		if(waitForPendingWebviewTasksToFinish) {
+			waitForPendingWebviewTasksToFinish();
+		}
+		return webViewFuture.join();
+	}
+
+	private void initializeWebView(ICoreWebView2 webView) {
+		webViewFuture.complete(webView);
+	}
+
+	private boolean isWebViewUnavailable() {
+		return webViewFuture.isCancelled();
+	}
+
+	private ICoreWebView2_2 getWebView_2(boolean waitForPendingWebviewTasksToFinish) {
+		if(waitForPendingWebviewTasksToFinish) {
+			waitForPendingWebviewTasksToFinish();
+		}
+		return webView_2Future.join();
+	}
+
+	private void initializeWebView_2(ICoreWebView2_2 webView_2) {
+		webView_2Future.complete(webView_2);
+	}
+
+	private void cancelWebView_2Initialization() {
+		webView_2Future.cancel(true);
+	}
+
+	private boolean isWebView_2Unavailable() {
+		return webView_2Future.isCancelled();
+	}
+
+	private ICoreWebView2_11 getWebView_11(boolean waitForPendingWebviewTasksToFinish) {
+		if(waitForPendingWebviewTasksToFinish) {
+			waitForPendingWebviewTasksToFinish();
+		}
+		return webView_11Future.join();
+	}
+
+	private void initializeWebView_11(ICoreWebView2_11 webView_11) {
+		webView_11Future.complete(webView_11);
+	}
+
+	private void cancelWebView_11Initialization() {
+		webView_11Future.cancel(true);
+	}
+
+	private boolean isWebView_11Unavailable() {
+		return webView_11Future.isCancelled();
+	}
+
+	/*
+	 * Schedule a given consumer with the given webViewFuture in the webViewTask
+	 * queue (using lastWebViewTask).
+	 * Note: This will be executed in the thread from
+	 * which it is called and must be therefore called from the UI thread.
+	 */
```

Comment:
Please adapt this documentation:
- there is no webViewFuture given and as a caller, you should know nithing about webViewTask and lastWebViewTask (implementation detail of this class)
- "the thread from which it is called" --> what is "it"? Is it this method or is it the action via the completion of the underlying future? From my understand, it can be both: in case the future is not completed yet, the completing thread will execute it. In case the future is already completed, it will be executed directly, i.e., within the calling thread. For the first case that the one who completes the future, it is difficult: how will I know when submitting this task who will complete the future? And is this an issue at all? The future will be completed by whomever calls `initializeWebView()`. This is called from the WebView component initialization callback, which, in turn, is executed via OS event processing, which, in turn, is always performed in the UI thread. So isn't the only relevant scenario here that the future is already completed and thus the task is executed _immediately_ in the calling thread? Then the documentation should state that it _might_ be executed immediately in the calling thread thus requiring to be in the UI thread in case anything within the action requires to be executed in the UI thread.

Reply:
Aah completely missed it. It's the old documentation. I'll change it.